### ISSUE TITLE
Error when SWRL rules have concrete values instead of arguments

### ISSUE THREAD

**Issue by Rezenders** [remaro-network/dlToPlanning#4]
I am getting the following error when transforming the suave ontology. 

```
Exception in thread "main" java.lang.ClassCastException: class uk.ac.manchester.cs.owl.owlapi.SWRLIndividualArgumentImpl cannot be cast to class org.semanticweb.owlapi.model.SWRLVariable (uk.ac.manchester.cs.owl.owlapi.SWRLIndividualArgumentImpl and org.semanticweb.owlapi.model.SWRLVariable are in unnamed module of loader 'app')
	at no.uio.tobiajoh.rules.RuleAssertionFactory.inferredRuleAssertion(RuleAssertionFactory.kt:39)
	at no.uio.tobiajoh.rules.DerivationRuleFactory.parseSWRLRule(DerivationRuleFactory.kt:72)
	at no.uio.tobiajoh.rules.DerivationRuleFactory.derivationRule(DerivationRuleFactory.kt:49)
	at no.uio.tobiajoh.OntologyTranslator.translateOWL(OntologyTranslator.kt:23)
	at no.uio.tobiajoh.Main.putTBoxIntoPDDL(Main.kt:59)
	at no.uio.tobiajoh.Main.run(Main.kt:43)
	at com.github.ajalt.clikt.parsers.Parser.finalizeAndRun(Parser.kt:348)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:218)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:42)
	at com.github.ajalt.clikt.core.CliktCommand.parse(CliktCommand.kt:457)
	at com.github.ajalt.clikt.core.CliktCommand.parse$default(CliktCommand.kt:454)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:474)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:481)
	at no.uio.tobiajoh.MainKt.main(Main.kt:71)

```

If I understood correctly, the error is due to some SWRL rules having concrete values instead of variables. For example: `tomasys:FunctionGrounding(?fg) ^ tomasys:solvesO(?fg, ?o) ^ tomasys:fg_status(?fg, "IN_ERROR_COMPONENT") ^ tomasys:Objective(?o) -> tomasys:o_status(?o, "IN_ERROR_COMPONENT")`

The error occurs in the following line when `it` is cast to `SWRLVariable`, which doesn't work since `it` is a `SWRLIndividualArgumentImpl`
https://github.com/remaro-network/dlToPlanning/blob/0816c93f8aa404ddd13524b6dd83992a2699521c/src/main/kotlin/no/uio/tobiajoh/rules/RuleAssertionFactory.kt#L37

The first problem to be solved is to think about how such rules should be coded in PDDL. Something like (I am really not sure about this hehe):
```
:derived (inferred-o_status ?o ?fg_status) 
		(exists (?fg ?fg_status ?o ?o_status2  )
 			(and
				(inferred-FunctionGrounding ?fg)
				(inferred-SolvesO ?fg ?o)
				(inferred-fg_status ?fg ?fg_status)
                                (in_error_component ?fg_status)
				(inferred-Objective ?o)
			)
		)
 	)
```

## Reproducing:
Switch to suave branch and run:
```./OWLToPDDL.sh --in=examples/suave/suave.pddl --owl=examples/suave/suave.owl --out=examples/suave/suave_created.pddl```

**Comment by tobiaswjohn** [comment]
I added some code to prevent the exception. However, we need indeed think about what the result should be in PDDL. For now, I parsed the arguments as constants to something like this:
```
(:derived (inferred-O_status ?o IN_ERROR_FR) 
		(exists (?fg)
 			(and
				(inferred-Objective ?o)
				(inferred-FunctionGrounding ?fg)
				(inferred-SolvesO ?fg ?o)
				(inferred-Fg_status ?fg IN_ERROR_FR)
			)
		)
 	)
)
```

This is exactly how the SWRL-rules in the ontology are. But a more general form (i.e. the one you provided) should also be doable, but is more a matter of changing the ontology.

Note: the current solution does not produce an exception, but the generated PDDL file might not work for planning, as the rules might introduce constants that are not defined anywhere else. As this is a matter of how / to which extend the ABox of the ontology is parsed, we need to decide this later. So this issue remains open for the time being.

**Comment by Rezenders** [comment]
I tested it, and it worked. It parsed the rule in the suave ontology to pddl, but not the rules that `suave.owl` inherithes from its upper (not sure if this is the correct term) ontologies.  
I guess this is a feature and not a bug, right? Should I move the rules from `tomasys.owl` to `suave.owl`?

> the current solution does not produce an exception, but the generated PDDL file might not work for planning, as the rules might introduce constants that are not defined anywhere else. As this is a matter of how / to which extend the ABox of the ontology is parsed, we need to decide this later. So this issue remains open for the time being.

What if we do the following:

1. Create a type for the `DataProperty`
2. Create a predicate for the `DataProperty` range
3. Create predicates for each value in the `DataProperty` range
4. Add to the problem file objects to represent each of the `DataProperty` range values
5. Init in the problem file the predicates created in the domain file

Example:
Domain file:
```
(:types
    fg-status
)

(:predicates
    (fg-status-range ?fg-status - fg-status)
    (fg-status-IN_ERROR_COMPONENT ?fg-status - fg-status)
    (fg-status-IN_ERROR_FR ?fg-status - fg-status)
    (fg-status-IN_ERROR_NFR ?fg-status - fg-status)
)

(:derived (inferred-Fg_status ?fg ?fg-status)
  (exists (?mqa ?fd ?fg-status ?eqa ?mqav ?eqav)
    (and
      (fg-status-range ?fg-status)
      (fg-status-IN_ERROR_COMPONENT ?fg-status)
      (inferred-FunctionGrounding ?fg)
      (inferred-QAvalue ?mqa)
      (inferred-HasQAvalue ?fg ?mqa)
      (inferred-IsQAtype ?mqa water_visibility)
      (inferred-FunctionDesign ?fd)
      (inferred-QAvalue ?eqa)
      (inferred-TypeFD ?fg ?fd)
      (inferred-HasQAestimation ?fd ?eqa)
      (inferred-IsQAtype ?eqa water_visibility)
      (inferred-HasValue ?mqa ?mqav)
      (inferred-HasValue ?eqa ?eqav)
      (inferred-LessThan ?mqav ?eqav)
    )
  )
)

``` 

Problem file:
```
    (:objects
      IN_ERROR_COMPONENT - fg-status
      IN_ERROR_FR - fg-status
      IN_ERROR_NFR - fg-status
    )

    (:init
      (fg-status-range IN_ERROR_COMPONENT)
      (fg-status-range IN_ERROR_FR)
      (fg-status-range IN_ERROR_NFR)

      (fg-status-IN_ERROR_COMPONENT IN_ERROR_COMPONENT)
      (fg-status-IN_ERROR_FR IN_ERROR_FR)
      (fg-status-IN_ERROR_NFR IN_ERROR_NFR)
    )
```

I think this would work for `DataProperty` with `boolean` range and `data range expression`. For `decimals` and `floats` we would need something else, perhaps fluents.

**Comment by tobiaswjohn** [comment]
should be fixed now with commit f85416c
